src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,274 @@
+        {
+            ThrowExceptionIfDisposed();
+
+            if (Read())
+            {
+                switch (_reader.TokenType)
+                {
+                    case JsonTokenType.String:
+                        return new[] { (string)_reader.GetString() };
+
+                    case JsonTokenType.StartArray:
+                        return ReadStringArrayAsReadOnlyListFromArrayStart();
+
+                    case JsonTokenType.StartObject:
+                        return null;
+                }
+            }
+
+            return null;
+        }
+
+        internal IReadOnlyList<string> ReadStringArrayAsReadOnlyListFromArrayStart()
+        {
+            ThrowExceptionIfDisposed();
+
+            List<string> strings = null;
+
+            while (Read() && _reader.TokenType != JsonTokenType.EndArray)
+            {
+                string value = _reader.ReadTokenAsString();
+
+                strings = strings ?? new List<string>();
+
+                strings.Add(value);
+            }
+
+            return (IReadOnlyList<string>)strings ?? Array.Empty<string>();
+        }
+
+        // This function is called when Read() returns false
+        private void GetMoreBytesFromStream()
+        {
+            int leftoverBytes = 0;
+            int bytesReadFromStream;
+
+            if (_reader.BytesConsumed < _buffer.Length)
+            {
+                // If the number of bytes consumed by the reader is less than the buffer size then we have leftover bytes that need to be shifted
+                var oldBuffer = _buffer;
+                ReadOnlySpan<byte> leftover = oldBuffer.AsSpan((int)_reader.BytesConsumed);
+
+                var returnOldBuffer = false;
+
+                // If the leftover bytes are the same as the buffer size then we are at capacity and need to double the buffer size
+                if (leftover.Length == _buffer.Length)
+                {
+                    returnOldBuffer = true;
+                    _buffer = ArrayPool<byte>.Shared.Rent(_buffer.Length * 2);
+                }
+
+                //Copy the leftover bytes to the beginning of the new buffer
+                leftover.CopyTo(_buffer);
+
+                // Read the rest of the bytes from the stream, keeping track of the number of bytes that need to be processed in the new buffer
+                leftoverBytes = leftover.Length;
+                bytesReadFromStream = Stream.Read(_buffer, leftover.Length, _buffer.Length - leftover.Length);
+                if (returnOldBuffer)
+                {
+                    ArrayPool<byte>.Shared.Return(oldBuffer);
+                }
+            }
+            else
+            {
+                bytesReadFromStream = Stream.Read(_buffer, 0, _buffer.Length);
+            }
+            _reader = new Utf8JsonReader(_buffer.AsSpan(0, leftoverBytes + bytesReadFromStream), isFinalBlock: leftoverBytes + bytesReadFromStream < _buffer.Length, _reader.CurrentState);
```

Comment:
I find this much easier to read. I'm really not a fan of expressions in method parameters, but I rarely see anyone else in the team complain about it, so take it with a grain of salt.

```suggestion
            var readerSpan = _buffer.AsSpan(0, leftoverBytes + bytesReadFromStream);
            var isFinalBlock = leftoverBytes + bytesReadFromStream < _buffer.Length
            _reader = new Utf8JsonReader(readerSpan, isFinalBlock, _reader.CurrentState);
```

However, as mentioned previously, [`Stream.Read`'s remarks says](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-8.0#system-io-stream-read(system-byte()-system-int32-system-int32))

> An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached.

Therefore, it's an unsafe assumption for `isFinalBlock` to be set when the buffer is not filled to capacity.